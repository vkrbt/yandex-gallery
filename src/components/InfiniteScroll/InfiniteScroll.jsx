import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '../Spinner/Spinner';
import { ScrollToTopButton } from '../ScrollToTopButton/ScrollToTopButton';

export class InfiniteScroll extends Component {
  constructor() {
    super();
    this.state = {
      loading: false,
    };
    this.handleScroll = this.handleScroll.bind(this);
    this.handleLoadingEnd = this.handleLoadingEnd.bind(this);
    this.handleGetNext = this.handleGetNext.bind(this);
  }

  componentDidMount() {
    this.handleGetNext();
    document.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll(e) {
    const windowHeight = window.innerHeight;
    const currentScroll = e.target.documentElement.scrollTop;
    if (!this.state.loading) {
      const maxScroll = e.target.documentElement.offsetHeight - windowHeight;
      if (currentScroll + windowHeight > maxScroll) {
        this.handleGetNext();
      }
    }
  }

  handleLoadingEnd() {
    this.setState({ loading: false });
  }

  async handleGetNext() {
    this.setState({ loading: true });
    try {
      await this.props.getNext();
    } finally {
      this.handleLoadingEnd();
    }
  }

  render() {
    return (
      <React.Fragment>
        {this.props.children}
        <Spinner
          loading={this.state.loading}
          action={this.handleGetNext}
          text={this.props.error && !this.state.loading ? 'Retry' : 'Load More'}
        />
        <ScrollToTopButton />
      </React.Fragment>
    );
  }
}

InfiniteScroll.propTypes = {
  getNext: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  error: PropTypes.bool.isRequired,
};
