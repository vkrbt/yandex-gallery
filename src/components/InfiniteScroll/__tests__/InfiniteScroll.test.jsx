import React from 'react';
import { shallow } from 'enzyme';
import { InfiniteScroll } from '../InfiniteScroll';

describe('InfiniteScroll', () => {
  it('should being rendered normally', () => {
    const props = {
      getNext: jest.fn(),
      children: <div>children</div>,
      error: false,
    };
    const wrapper = shallow(<InfiniteScroll {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
