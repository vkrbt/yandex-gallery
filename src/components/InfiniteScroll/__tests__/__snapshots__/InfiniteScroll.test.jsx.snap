// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll should being rendered normally 1`] = `
<React.Fragment>
  <div>
    children
  </div>
  <Component
    action={[Function]}
    loading={true}
    text="Load More"
  />
  <ScrollToTopButton />
</React.Fragment>
`;
